sw/host/opentitanlib/src/transport/hyperdebug/mod.rs

```diff
@@ -0,0 +1,320 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+
+use std::cell::RefCell;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::rc::Rc;
+
+use thiserror::Error;
+
+use crate::io::gpio::GpioPin;
+use crate::io::spi::Target;
+use crate::io::uart::Uart;
+use crate::transport::{Capabilities, Capability, Transport, TransportError};
+use crate::util::usb::UsbBackend;
+
+pub mod gpio;
+pub mod uart;
+pub mod spi;
+
+/// Implementation of the Transport trait for HyperDebug based on the
+/// Nucleo-L552ZE-Q.
+pub struct Hyperdebug {
+    spi_names: HashMap<String, u8>,
+    spi_interface: BulkInterface,
+    uart_ttys: HashMap<String, PathBuf>,
+    inner: Rc<Inner>,
+}
+
+/// Index of a single USB "interface", with its associated IN and OUT
+/// endpoints.  Used to instantiate e.g. SPI trait.
+#[derive(Copy, Clone)]
+pub struct BulkInterface {
+    interface: u8,
+    in_endpoint: u8,
+    out_endpoint: u8,
+}
+
+/// Internal state of the Hyperdebug struct, this struct is reference counted such that Gpio,
+/// Spi and Uart sub-structs can all refer to this shared data, which is guaranteed to live on,
+/// even if the caller lets the outer Hyperdebug struct run out of scope.
+pub struct Inner {
```

Comment:
Could you move this to just before the corresponding impl block?

Reply:
Done